### PR TITLE
Add some documentation + fix two clippy issues

### PR DESCRIPTION
**Pull request by oyvindln**
Clippy found more issues, but they were deep in macros, so I wasn't sure about how or if to go about fixing them.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -6,114 +6,325 @@
...
-    0b0010_0111, 0b1010_0111, 0b0110_0111, 0b1110_0111,
-    0b0001_0111, 0b1001_0111, 0b0101_0111, 0b1101_0111,
-    0b0011_0111, 0b1011_0111, 0b0111_0111, 0b1111_0111,
-
-    0b0000_1111, 0b1000_1111, 0b0100_1111, 0b1100_1111,
-    0b0010_1111, 0b1010_1111, 0b0110_1111, 0b1110_1111,
-    0b0001_1111, 0b1001_1111, 0b0101_1111, 0b1101_1111,
-    0b0011_1111, 0b1011_1111, 0b0111_1111, 0b1111_1111
-];
+static BIT_REV_U8: [u8; 256] = [0b0000_0000,
```
Comment (eddyb) [inline src/lib.rs]: Please do not change formatting where manual arrangement was used for semantic purposes (this is a lookup table, not a list).

File: src/lib.rs [hunk 2]
```diff
@@ -428,11 +643,17 @@ impl DynHuffman16 {
...
-    Uncompressed(/* len */ u16),
-    CheckCRC
+    LenDist((BitsNext, BitState),
+            // len
+            u16,
+            // dist
+            u16),
+    Uncompressed(// len
+                 u16),
+    CheckCRC,
```
Comment (eddyb) [inline src/lib.rs]: The changes here really don't work out.

File: src/lib.rs [hunk 3]
```diff
@@ -441,11 +662,29 @@ enum BitsNext {
...
+                            // clens
+                            Box<[u8; 19]>),
     BlockDynCodeLengths(CodeLengthReader),
-    BlockDyn(/* lit/len */ DynHuffman16, /* dist */ DynHuffman16, /* prev_len */ u16)
+    BlockDyn(// lit/len
+             DynHuffman16,
+             // dist
+             DynHuffman16,
+             // prev_len
+             u16),
```
Comment (eddyb) [inline src/lib.rs]: Same.

File: src/lib.rs [hunk 4]
```diff
@@ -523,16 +763,16 @@ impl InflateStream {
...
         };
-        
+
         if self.buffer.len() < pos_end as usize {
             unsafe {
                 self.buffer.set_len(pos_end as usize);
             }
         }
-        
+
```
Comment (eddyb) [inline src/lib.rs]: Huh, weird, my editor is configured to strip these, I wonder how they got here.

File: src/lib.rs [hunk 5]
```diff
@@ -579,11 +818,8 @@ impl InflateStream {
         macro_rules! run_len_dist (($len:expr, $dist:expr => ($bytes:expr, $next:expr, $state:expr)) => ({
             let dist = $dist;
             let left = try!(self.run_len_dist($len, dist));
-            match left {
-                Some(len) => {
-                    return ok_bytes!($bytes, LenDist(($next, $state), len, dist));
-                }
-                None => {}
+            if let Some(len) = left {
```
Comment (eddyb) [inline src/lib.rs]: Niiice!

File: src/lib.rs [hunk 6]
```diff
@@ -592,15 +828,17 @@ impl InflateStream {
                 let (method, info) = (b & 0xF, b >> 4);
                 debug!("ZLIB CM=0x{:x} CINFO=0x{:x}", method, info);
                 match method {
-                    8 => {/* DEFLATE */}
-                    _ => return Err(format!("unknown ZLIB method CM=0x{:x}", method))
+                    8 => {
+                        // DEFLATE
+                    }
```
Comment (eddyb) [inline src/lib.rs]: Seems unnecessary. In general I disagree with using line comments for things that aren't on their own line.

File: src/lib.rs [hunk 7]
```diff
@@ -783,7 +1025,8 @@ impl InflateStream {
                             if (stream.state.v & 0b11) != 0b11 {
                                 save = stream.clone();
                                 // FIXME(eddyb) use a specialized rev LUT with addend.
-                                let code = BIT_REV_U8[(stream.take(8).unwrap()) as usize] - 0b0011_0000;
+                                let code = BIT_REV_U8[(stream.take(8).unwrap()) as usize] -
+                                           0b0011_0000;
```
Comment (eddyb) [inline src/lib.rs]: At least align just after `[`.

File: src/lib.rs [hunk 8]
```diff
@@ -792,13 +1035,15 @@ impl InflateStream {
                             if (stream.state.v & 0b11100) == 0b00000 {
                                 save = stream.clone();
                                 // FIXME(eddyb) use a 3-bit rev LUT or match the huffman code directly.
-                                let code = 24 + (BIT_REV_U8[stream.take(8).unwrap() as usize] - 0b11000000);
+                                let code = 24 +
+                                           (BIT_REV_U8[stream.take(8).unwrap() as usize] -
+                                            0b11000000);
```
Comment (eddyb) [inline src/lib.rs]: Same.

File: src/lib.rs [hunk 9]
```diff
@@ -818,11 +1063,13 @@ impl InflateStream {
...
-                            None => return ok!(BlockDynClenCodeLengths(hlit, hdist, hclen, i, clens))
-                        };
-                        clens[[16, 17, 18, 0, 8, 7, 9, 6, 10, 5, 11, 4, 12, 3, 13, 2, 14, 1, 15][i as usize]] = v;
+                        let v =
+                            match stream.take(3) {
+                                Some(v) => v,
+                                None => return ok!(BlockDynClenCodeLengths(hlit, hdist, hclen, i, clens)),
+                            };
+                        clens[[16, 17, 18, 0, 8, 7, 9, 6, 10, 5, 11, 4, 12, 3, 13, 2, 14, 1,
+                               15][i as usize]] = v;
```
Comment (eddyb) [inline src/lib.rs]: Ugh, I really don't like random splitting like this, it's a lookup table.

**Comment by eddyb** [comment]
Btw you can make a list of clippy issues and I can at least tell you how many are aplicable given the intent of the code to compactly express the inflate algorithm.

**Comment by oyvindln** [comment]
The formatting is just the rustfmt defaults, and yeah it sometimes messes up things, I'll go through and undo the things you suggested for now. 

As for the clippy issues, I put the full output from clippy here: https://gist.github.com/oyvindln/f54bbbf1041a0e53feb56810143885d2

It's very long, as two of the issues are found in macros, which are invoked a lot.

**Comment by eddyb** [comment]
Oh, you have to allow `eq_op` because it's triggeres by macro unrolling (we might be able to side-step it by complicating the macro but I'm not sure).
As for `eval_order_dependence`, it just looks wrong. cc @Manishearth, who came up with the idea that evaluation order is not defined in Rust? I mean, other than assignment expressions.

**Comment by Manishearth** [comment]
Yeah, you should be allowing lints when you think they're ok. Clippy will have false positives in the context of macros.

**Comment by oyvindln** [comment]
Okay, I reverted the formatting changes as you requested. I also mirrored the documentation to the README file.

**Comment by eddyb** [comment]
Thanks!

**Pull request merged**